text/0081-sourcemap-debugid.md

```diff
@@ -0,0 +1,362 @@
+belong to a different release, due to caching or other reasons.
+
+Using a `DebugId`, which uniquely associates the JavaScript file and its corresponding SourceMap, should make source-mapping
+a lot more reliable.
+
+# Supporting Data
+
+TODO: please fill in the gaps here!
+
+Sentry has used the `release + dist` solution for quite some time and found it inadequate.
+A lot of events are not being resolved correctly due to these mismatches, and problems with source-mapping are very
+common in customer-support interactions.
+
+On the other hand, using a `DebugId` for symbolication of Native crashes and stack traces is working reliably both in
+Sentry and in the wider native ecosystem. The Native and C# community has the concept of _Symbol Servers_, which can
+serve any debug file based on its `DebugId`, which allows reliable symbolication for any release, at any point in time.
+
+# Options Considered
+
+To make `DebugId` work, we need to generate one, and associate it to both the JavaScript file, and its corresponding
+SourceMap.
+
+## The `DebugId` format
+
+The `DebugId` should have the same format as a standard UUI, specifically:
+It should be a 128 bit (16 byte), formatted to a string using base-16 hex encoding like so:
+
+`XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX`
+
+## How to generate `DebugId`s?
+
+There is two options of choosing a `DebugId`: Making it completely random, or make it reproducible by deriving it from
+a content hash.
+
+### Based on JavaScript Content-hash
+
+This creates a new `DebugId` by hashing the contents of the JavaScript file.
+
+**pros**
+
+- Is fully reproducible. The same JavaScript file will always have the same `DebugId`.
+- Works well with existing caching solutions.
+
+**cons**
+
+- Increases overhead in server-side SourceMap processing, as one file can potentially be included in multiple _bundles_.
+  See [_What is an `ArtifactBundle`_](#what-is-an-artifactbundle) below.
+- A difference in a source file might not be reflected in the JavaScript file. An example of this might be changes to
+  whitespace, comments, or code that was dead-code-eliminated by bundlers.
+
+### Based on SourceMap Content-hash
+
+This creates a new `DebugId` by hashing the contents of the SourceMap file.
+
+**pros**
+
+- Generates a new `DebugId` for changes to source files that would otherwise not lead to changes in the JavaScript file.
+
+**cons**
+
+- Does lead to slightly more cache invalidation.
+
+### Random `DebugId`
+
+This option would create a new random `DebugId` for each file, on each build.
+
+**pros**
+
+- Simpler server-side SourceMap processing, as one `DebugId` is only included in a single _bundle_, and that one bundle
+  can serve multiple stack frames for multiple files of the same build.
+
+**cons**
+
+- Completely breaks the concept of _caching_, as every file is unique for every build.
+
+## How to inject the `DebugId` into the JavaScript file?
```

Comment:
Do we actually need to do this? for example, the `proguard` cmd takes a `uuid` if required and makes the association.
https://github.com/getsentry/sentry-android-gradle-plugin/blob/a8fc82a9d45fdc0208b276fa18eb377bcaea440f/plugin-build/src/main/kotlin/io/sentry/android/gradle/tasks/SentryUploadProguardMappingsTask.kt#L85-L86

Reply:
Well, the UUID of the proguard mapping is embedded into the shipped app though, so the SDK can transmit that with every event.
This is the same concept, just that doing it for JS makes it a ton more complex.

Reply:
We also defined a process by which the proguard file comes to a deterministic debug id through generating a SHA1 hash from the content hash.

Reply:
We just need to be sure that the proguard file (mapping) is also deterministic, the proguard tooling changed multiple times (now R8), Java to Kotlin, etc, better to confirm, I asked Android folks to test it out with the recent tooling at least.

Reply:
I did a quick test for R8 + Android, rebuilding the same source from scratch multiple times and consistenly received the same mapping file:
````
% rm -rf build app/build
% ./gradlew :app:assembleRelease --no-build-cache --no-daemon
% md5 app/build/outputs/mapping/release/mapping.txt 
MD5 (app/build/outputs/mapping/release/mapping.txt) = e2ee64df9b51983445ffac36a4da455c
````
But this should be taken with a grain of salt, e.g. f-droid reports that in some cases R8 is not deterministic: https://f-droid.org/docs/Reproducible_Builds/#r8-optimizer

Reply:
Side note: I don't think we're using this right now but current versions of R8 generate a SHA-256 hash of the mapping payload and put it into the header of the mapping.txt file
````
head -n 10 app/build/outputs/mapping/release/mapping.txt
# compiler: R8
# compiler_version: 3.3.83
# min_api: 24
# common_typos_disable
# {"id":"com.android.tools.r8.mapping","version":"2.0"}
# pg_map_id: 0f71627
# pg_map_hash: SHA-256 0f71627e8e144f6a30bf2c32bef474bd7ab4f3173094061774e4c0722ebc8772
android.app.AppComponentFactory -> android.app.AppComponentFactory:
# {"id":"com.android.tools.r8.synthesized"}
    void <clinit>() -> <clinit>
````

Reply:
Ah nice, we might even use that if it is also being exposed anywhere inside of apps.

The concept of file identity is definitely also a concern for proguard / java and other ecosystems, however I would like to focus discussions in this RFC to JS/SourceMaps :-)